ghcide/src/Development/IDE/Core/Shake.hs

```diff
@@ -931,20 +941,25 @@ defineEarlyCutoff op = addBuiltinRule noLint noIdentity $ \(Q (key, file)) (old
                 (bs, (diags, res)) <- actionCatch
                     (do v <- op key file; liftIO $ evaluate $ force v) $
                     \(e :: SomeException) -> pure (Nothing, ([ideErrorText file $ T.pack $ show e | not $ isBadDependency e],Nothing))
-                modTime <- liftIO $ (currentValue . fst =<<) <$> getValues state GetModificationTime file
```

Comment:
There is a race here, we shouldn't assign the latest version of the file as the version, rather we should use the `Token` corresponding to the version that is in the store.

However, this is a bit tricky since `getVirtualFile` returns just a `TextDocumentVersion` and not a `Token`. We also can't maintain a map of `Token`s to . To fix this, we would need to explicitly synchronize the lsp VFS with the document change notifications we define handlers for. At that point, we would have a token available, and we can return that along with the virtual file. This means that the type of `getVirtualFile` would change to `NormalizedUri -> IO (Maybe (VirtualFile, Token))`. 

I believe this bit was already somewhat racy(https://github.com/haskell/haskell-language-server/issues/1172), since `getVirtualFile` would always return the latest version of the document according to lsp, which might be ahead of the 'snapshot'  of the VFS we assume in the Shake sessioin.

We should maintain a separate snapshot of the VFS which is updated (via `LSP.getVirtualFile`) on `documentChange` events. This store can also keep track of the `Tokens` corresponding to the versions of the file in the VFS.